lib/ecto/repo/model.ex

```diff
@@ -14,7 +14,7 @@ defmodule Ecto.Repo.Model do
     struct = struct_from_changeset!(changeset)
     model  = struct.__struct__
     fields = model.__schema__(:fields)
-    source = model.__schema__(:source)
+    source = struct.__meta__.source || model.__schema__(:source)
```

Comment:
Let's *always* read the source from `__meta__.source`. If it is nil, it should be a bug.